src/CrawlQueue/ArrayCrawlQueue.php

```diff
@@ -0,0 +1,102 @@
+<?php
+
+namespace Spatie\Crawler\CrawlQueue;
+
+use TypeError;
+use Spatie\Crawler\CrawlUrl;
+use Psr\Http\Message\UriInterface;
+use Spatie\Crawler\Exception\UrlNotFoundByIndex;
+
+/**
+ * Crawl queue implemented with arrays.
+ */
+class ArrayCrawlQueue implements CrawlQueue
+{
+    /**
+     * All known URLs, indexed by URL string.
+     *
+     * @var CrawlUrl[]
+     */
+    protected $urls = [];
+
+    /**
+     * Pending URLs, indexed by URL string.
+     *
+     * @var CrawlUrl[]
+     */
+    protected $pendingUrls = [];
+
+    public function add(CrawlUrl $url) : CrawlQueue
+    {
+        $urlString = (string) $url->url;
+
+        if (! isset($this->urls[$urlString])) {
+            $url->setId($urlString);
+
+            $this->urls[$urlString] = $url;
+            $this->pendingUrls[$urlString] = $url;
+        }
+
+        return $this;
+    }
+
+    public function hasPendingUrls() : bool
+    {
+        return (bool) $this->pendingUrls;
+    }
+
+    public function getUrlById($id) : CrawlUrl
+    {
+        if (! isset($this->urls[$id])) {
+            throw new UrlNotFoundByIndex("Crawl url $id not found in collection.");
+        }
+
+        return $this->urls[$id];
+    }
+
+    public function hasAlreadyBeenProcessed(CrawlUrl $url) : bool
+    {
+        $url = (string) $url->url;
+
+        return ! isset($this->pendingUrls[$url]) && isset($this->pendingUrls[$url]);
```

Comment:
Maybe it's because I don't have had coffee yet, but I don't understand this line. Could you explain what you're checking here?

Reply:
Added some tests around this and fixed this on `master`

Reply:
- fix line 61: 
```return ! isset($this->pendingUrls[$url]) && isset($this->processedUrls[$url]);```

Reply:
Looks like you figured it out in 9078d286aac9b96f9ba4ed2fc66d82a3237c495f ! This line is the strict translation of:

https://github.com/spatie/crawler/blob/287d8525b613d7301e411f4a22df952aba64df63/src/CrawlQueue/CollectionCrawlQueue.php#L71

But less formatted for readability.

Reply:
isset($this->pendingUrls[$url]) is repeated twice? doesn't this return false always?

Reply:
OMG I'm just seing it now, someone had to put it twice right under my eyes before I could see it. Thanks for pointing this out, it should have been `&& isset($this->urls[$url])`. This has been fixed in 9078d286aac9b96f9ba4ed2fc66d82a3237c495f anyway!

Reply:
I just wanna help out... cool thing is fixed